jctools-experimental/src/main/java/org/jctools/stacks/WFIStack.java

```diff
@@ -0,0 +1,373 @@
+     * The returned iterable is a complete and atomic snapshot of the stack contents at the time of the method call.
+     * Racing callers will not see any overlap between their returned iterables.
+     * The iterable is re-entrant, so {@link Iterable#iterator()} can be called on it multiple times, and it will
+     * always produce the same result.
+     * <p>
+     * This operation is wait-free and has constant time-complexity.
+     *
+     * @return An iterable of all the nodes in the stack, in Last In, First Out order.
+     */
+    public Iterable<T> replaceAll(T node)
+    {
+        checkPush(node);
+        node.count = 1;
+        node.next = END;
+        return new NodeIterable<T>(xchgHead(node));
+    }
+
+
+    /**
+     * Replaces and returns all nodes in the stack with the given node, and return an iterable of the removed nodes.
+     * The returned iterable is a complete and atomic snapshot of the stack contents at the time of the method call.
+     * Racing callers will not see any overlap between their returned iterables.
+     * The iterable is re-entrant, so {@link Iterable#iterator()} can be called on it multiple times, and it will
+     * always produce the same result.
+     * <p>
+     * This operation is wait-free and has time-complexity linear to the number of elements in the returned iterable.
+     *
+     * @return An iterable of all the nodes in the stack, in First In, First Out order.
+     */
+    public Iterable<T> replaceAllFifo(T node)
+    {
+        checkPush(node);
+        node.count = 1;
+        node.next = END;
+        return new NodeIterable<T>(xchgHead(node)).reverse();
+    }
+
+    /**
+     * Pop, or remove, the most recently pushed node, or return {@code null} if the stack is empty.
+     * <p>
+     * This operation is non-blocking, but not wait-free. It has constant time-complexity.
+     *
+     * @return The most recently pushed node, if any, or {@code null}.
+     */
+    public T pop()
+    {
+        Node candidate, successor;
+        do
+        {
+            candidate = head;
+            if (candidate == END)
+            {
+                return null;
+            }
+            successor = candidate.next();
+        }
+        while (!casHead(candidate, successor));
+        return candidate.self();
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return new NodeIterator<T>(head);
+    }
+
+    private void checkPush(T node)
+    {
+        assert node.next == null : "WFIStack.Nodes cannot be reused.";
+    }
+
+    private T xchgHead(Node node)
+    {
+        return (T) UnsafeAccess.UNSAFE.getAndSetObject(this, HEAD_OFFSET, node);
+    }
+
+    private boolean casHead(Node expected, Node update)
+    {
+        return UnsafeAccess.UNSAFE.compareAndSwapObject(this, HEAD_OFFSET, expected, update);
+    }
+}
+
+@SuppressWarnings("unused")
+abstract class PadBefore
```

Comment:
Please reorder and rename padding classes as per other classes in JCTools. i.e. padding in order of inheritance, padding fiends copy and pasted from existing code(helps searching)